### DISCUSSION TITLE
Trouble Hiding NoData Value

### DISCUSSION THREAD

**Discussion by dseminara-pivot** [developmentseed/titiler#643] [Q&A] [answered]
Hello there!

I just started experimenting with titiler yesterday and have made great progress in a rather short amount of time (deploying it to the cloud, rendering tiles in a variety of web mapping libraries, etc)! However, I can't seem to generate tiles that respect our `nodata` value of `-10000` and I'm hoping this is something as simple as me still being new to titiler.

In QGIS, it seems to automatically 'know' this value and drops it from view. Similarly, when using the same COG image outside of Titiler and directly from the source - I'm able to tell OpenLayers what the value is and have it appropriately remove it from the scene.

I've also tried adding `&nodata=-10000` to the end of the url and have used the swagger docs to generate it for me as well - but it doesn't seem to respect that param.

Attached is a `gdalinfo` output (minus location info like Origin and Corner Coordinates) and a small snippet of what I'm referring to. I saw that `rio faux` could be provided as an option, but unfortunately I can't provide any spatial location to where this image is from.

```shell
> gdalinfo out.tif
Driver: GTiff/GeoTIFF
Files: out.tif
Size is 11557, 10545
Coordinate System is:
GEOGCS["WGS 84",
    DATUM["WGS_1984",
        SPHEROID["WGS 84",6378137,298.257223563,
            AUTHORITY["EPSG","7030"]],
        AUTHORITY["EPSG","6326"]],
    PRIMEM["Greenwich",0],
    UNIT["degree",0.0174532925199433],
    AUTHORITY["EPSG","4326"]]
Pixel Size = (0.000000141738231,-0.000000141738231)
Metadata:
  AREA_OR_POINT=Area
  OVR_RESAMPLING_ALG=NEAREST
Image Structure Metadata:
  COMPRESSION=DEFLATE
  INTERLEAVE=PIXEL
Band 1 Block=512x512 Type=Float32, ColorInterp=Blue
  Description = Band1
  NoData Value=-10000
  Overviews: 5779x5273, 2890x2637, 1445x1319, 723x660, 362x330
Band 2 Block=512x512 Type=Float32, ColorInterp=Green
  Description = Band2
  NoData Value=-10000
  Overviews: 5779x5273, 2890x2637, 1445x1319, 723x660, 362x330
Band 3 Block=512x512 Type=Float32, ColorInterp=Red
  Description = Band3
  NoData Value=-10000
  Overviews: 5779x5273, 2890x2637, 1445x1319, 723x660, 362x330
Band 4 Block=512x512 Type=Float32, ColorInterp=Undefined
  Description = Band4
  NoData Value=-10000
  Overviews: 5779x5273, 2890x2637, 1445x1319, 723x660, 362x330
Band 5 Block=512x512 Type=Float32, ColorInterp=Undefined
  Description = Band5
  NoData Value=-10000
  Overviews: 5779x5273, 2890x2637, 1445x1319, 723x660, 362x330
Band 6 Block=512x512 Type=Float32, ColorInterp=Undefined
  Description = Band6
  NoData Value=-10000
  Overviews: 5779x5273, 2890x2637, 1445x1319, 723x660, 362x330
Band 7 Block=512x512 Type=Float32, ColorInterp=Undefined
  Description = Band7
  NoData Value=-10000
  Overviews: 5779x5273, 2890x2637, 1445x1319, 723x660, 362x330
```

**Tile Request in the browser:**
![browser-tile-request](https://github.com/developmentseed/titiler/assets/117767782/92a4fc08-65b0-42ff-a606-5286dd0c9115)

**Using /cog/preview**
![cog-preview](https://github.com/developmentseed/titiler/assets/117767782/537ef4fa-66ea-462e-b328-2db61086b2aa)

**Comment by vincentsarago** [comment]
👋 @dseminara-pivot 

Without access to the file it will be pretty hard for me to help. 

If there is an issue, it will be in the rio-tiler side anyway. TO debug this you could just try to play with rio-tiler 

```python
from rio_tiler.io import Reader

with Reader("my.tif") as src:
    im = src.preview(indexes=1)  # Read the first band
    
    data = im.data
    mask = im.mask  #<-- this is the part you want to check to make sure the nodata is used to create the mask
 ``` 
 
 Note: we are currently working on the next version of rio-tiler which change a lot of things for the mask creation (switch from a per-dataset to a per-band mask) and also fixing some issue when working with GCPS referenced files.

**Reply by dseminara-pivot** [reply]
Thanks for the quick reply @vincentsarago! ..and totally understand the difficulties with assistance without access to our image dataset. Does perhaps the rescale param round? Our pixel values are incredibly low with `rescale=0,0.006`. I've tried modifying the min value to be greater than 0 but below our max of `0.006` but get the same result.

Looking at the file within `Reader`, i see `data` with the matching `-10000.0` value for what should be expect to be  nodata and within `mask` a value of `0`

**Reply by vincentsarago** [reply]
the mask value should be either 0 or 255 (0 mean masked, 255 means no masked)

**Comment by dseminara-pivot** [comment]
Perhaps asking this another way - in hopes I'm not misunderstanding what the mask does. Does/can the mask provide a transparent 'value' similar to when a tile is at edge of the actual image. Basically, I'm hoping that the black area which represents `nodata` can transparent as well.

![transparent-mask-img](https://github.com/developmentseed/titiler/assets/117767782/34b3b9d6-b79e-4e61-b72c-d5f114b4415c)

**Reply by vincentsarago** [reply]
Yes it should be transparent as well. 

if you cannot share the data publicly but can directly to me, please feel free to contact me at vincent@developmentseed.org

**Reply by vincentsarago** [reply, marked as answer]
after inspecting the data I think I know what's going on! 

The band 6 and 7 do not have the same `nodata` mask. Their value is set to `0` while the other bands are at `-10000`

<img width="402" alt="Screenshot 2023-05-12 at 12 52 26 AM" src="https://github.com/developmentseed/titiler/assets/10407788/e39a09bb-201d-47aa-b7a3-33ae757556dd">

Sadly the current version of rio-tiler use a `per-dataset` mask, meaning that it will construct a mask using the `sum` of all the mask, and because the band 6 and 7 says it shouldn't mask the area on the right of the image, the combined mask return 255 instead of 0. 

https://github.com/cogeotiff/rio-tiler/blob/main/rio_tiler/reader.py#L197-L202

Next version of rio-tiler will use `per-band mask` so it should better handle this case (ref: https://github.com/cogeotiff/rio-tiler/issues/579) 

https://github.com/cogeotiff/rio-tiler/blob/30dc83cdc3154ac2820363d33f1ef6ced8cb8d25/rio_tiler/reader.py#L222-L233

### Possible fix 

On your side, while we wait for new rio-tiler/titiler, you could edit your file to have a kinda unified nodata area (editing band 6 and 7 to have the same values as band 1 where it's -10000)

**Reply by dseminara-pivot** [reply]
Thanks Vincent. That makes sense...our last two bands are based on derived values which probably explains the mismatch. Appreciate you investigating.
